drivers/apc_modbus.c

```diff
@@ -0,0 +1,1328 @@
+
+	Free_ReportDesc(hid_desc);
+
+usb_callback_exit:
+	dstate_setinfo("ups.vendorid", "%04x", device->vid);
+	dstate_setinfo("ups.productid", "%04x", device->pid);
+
+	return 0;
+}
+#endif /* defined NUT_MODBUS_HAS_USB */
+
+static int _apc_modbus_parse_host_port(const char *input, char *host, size_t host_buf_size, char *port, size_t port_buf_size, const uint16_t default_port) {
+	const char *start = input;
+	const char *end = input;
+	int port_int;
+	size_t host_size, port_size;
+
+	if (*start == '[') {
+		start++;
+		end = strchr(start, ']');
+		if (!end) {
+			upslogx(LOG_ERR, "%s: Invalid IPv6 notation", __func__);
+			return 0;
+		}
+	} else {
+		end = strchr(start, ':');
+	}
+
+	if (!end) {
+		/* Port is missing, use the default port */
+		strncpy(host, start, host_buf_size);
+		snprintf(port, port_buf_size, "%u", default_port);
+		return 0;
+	}
+
+	host_size = (size_t)(end - start);
+	port_size = strlen(end + 1);
+
+	if (host_size >= host_buf_size || port_size >= port_buf_size) {
+		upslogx(LOG_ERR, "%s: Buffer size too small", __func__);
+		return 0;
+	}
+
+	strncpy(host, start, host_size);
+	strncpy(port, end + 1, port_size);
+
+	port_int = atoi(port);
+	if (port_int < 0 || port_int > 65535) {
+		upslogx(LOG_ERR, "%s: Port out of range", __func__);
+		return 0;
+	}
+
+	return 1;
+}
+
+void upsdrv_initups(void)
+{
+	int r;
+#if defined NUT_MODBUS_HAS_USB
+	char *regex_array[USBMATCHER_REGEXP_ARRAY_LIMIT];
+	int has_nonzero_regex;
+	size_t i;
+#endif /* defined NUT_MODBUS_HAS_USB */
+	char *val;
+	int rtu_baudrate;
+	char rtu_parity;
+	int rtu_databits;
+	int rtu_stopbits;
+	int slaveid;
+	uint32_t response_timeout_ms;
+	char tcp_host[256];
+	char tcp_port[6];
+
+	val = getval("porttype");
+
+	if (val == NULL) {
+#if defined NUT_MODBUS_HAS_USB
+		val = "usb";
+#else
+		val = "serial";
+#endif /* defined NUT_MODBUS_HAS_USB */
+	}
+
+	is_open = 0;
+
+#if defined NUT_MODBUS_HAS_USB
+	is_usb = 0;
+#endif /* defined NUT_MODBUS_HAS_USB */
+
+	if (!strcasecmp(val, "usb")) {
+#if defined NUT_MODBUS_HAS_USB
+		is_usb = 1;
+
+		/* We default to USB, this is the most common connection type and does not require additional
+		 * parameter, so we can auto-detect these. */
+
+		warn_if_bad_usb_port_filename(device_path);
+
+		regex_array[0] = getval("vendorid");
+		regex_array[1] = getval("productid");
+		regex_array[2] = getval("vendor");
+		regex_array[3] = getval("product");
+		regex_array[4] = getval("serial");
+		regex_array[5] = getval("bus");
+		regex_array[6] = getval("device");
+#  if (defined WITH_USB_BUSPORT) && (WITH_USB_BUSPORT)
+		regex_array[7] = getval("busport");
+#  else
+		if (getval("busport")) {
+			upslogx(LOG_WARNING, "\"busport\" is configured for the device, but is not actually handled by current build combination of NUT and libusb (ignored)");
+		}
+#  endif
+
+		has_nonzero_regex = 0;
+		for (i = 0; i < sizeof(regex_array) / sizeof(regex_array[0]); i++) {
```

Comment:
Isn't `sizeof(array)` a count of its elements (vs. a byte size)? Maybe explicit `USBMATCHER_REGEXP_ARRAY_LIMIT` is reasonable here...

Reply:
No, `sizeof` is the byte size, which is why it is divided by the size of the first element. I'll change it to the macro if you want to, but I personally like using the array size more.

Reply:
I like having a `#define COUNTOF(x) (sizeof(x) / sizeof(x[0]))` in my codebases

Reply:
My reasoning for using sizeof is that things can get out of sync if you don't. If somone were to change the array size without adjusting the macro, the range of the loop would not match. Using `sizeof` it will always match.

Reply:
Yeah, I know the benefits just was a bit confused jumping too many languages a week :\

For the COUNTOF, I guess it could be added... perhaps to `common.h`?.. Wondering if a nullness check might be worthwhile too :)

Reply:
It is a stack local variable, so NULL checks shouldn't be needed :) We can add `COUNTOF`, but I'd like to just get this merged. I'll make another PR for the commands soon, we can address this there.

Reply:
But maybe as another PR